index.rst

```diff
@@ -10,16 +10,200 @@
+For example, preventing a retry or a delayed processing run from using any APDB entries added after the "correct" time can lead to two visits creating DiaObjects at the same position because each is required to ignore the other.
+More complex strategies using validity ranges or other tools can avoid such paradoxes, but may lead to more subtle bugs.
+
+The simplest way to keep a consistent association order when recovering from processing failures is to allow all runs to use the state of the APDB at the final processing time.
+If we (and science users) think of out-of-order visits as precoveries, then there shouldn't be any confusion over the processing order not being strictly chronological.
+
+.. _consistency:
+
+Inconsistent Output
+===================
+
+As noted in :ref:`general-priorities`, the most important pipeline outputs are, in order, the APDB, the alert stream, and the central repository and metrics.
+As of May 2023, this is also the order in which the pipeline produces outputs.
+If pipeline processing fails in its late stages, these outputs may be inconsistent with each other; for example, the APDB may contain DiaSources for which alerts were never sent.
+
+The alert stream can be easily restored from the APDB, if we have a record of which alerts were sent or can tolerate duplicates.
+The reverse conversion is unsafe, because injecting associations after other visits have been processed could lead to contradictory source association histories (see :ref:`association`).
+
+The central repository can itself acquire inconsistencies in two ways.
+First, we will try to transfer the outputs from failed (i.e., incomplete) pipelines, as these may help in diagnosing the problem.
+Such a strategy is safe so long as no code assumes that the existence of one dataset implies the existence of another.
+Second, the central repository sync itself may fail, leaving an undefined subset of datasets transferred.
+Again, the immediate risk is that something might assume the repository contains a self-consistent set of outputs; in the longer term, the datasets can be regenerated through daytime processing.
+
+
+.. _bad-data:
+
+Corrupted Pipeline Outputs
+==========================
+
+It's possible that some processing errors will allow the pipeline to run to completion while producing large numbers of invalid sources.
+Such sources will clutter the alert stream with false positives, and may confuse source association on later visits.
+
+The Prompt Processing framework does not itself have any way to detect nonsense output.
+However, the Alert Production team is incorporating "circuit breaker" checks into the pipeline; see `DM-35645`_ and its follow-up issues.
+These checks will escalate suspicious outputs into pipeline failures, which can be handled as described above.
+As of May 2023, the proposed checks focus on poor-quality raw inputs; there are no checks specifically guarding DiaSouce detection or the APDB.
```

Comment:
```suggestion
As of May 2023, the proposed checks focus on poor-quality raw inputs; there are no checks specifically guarding DiaSource detection or the APDB.
```

Reply:
@ebellm, is this statement still true? [DM-35645](https://jira.lsstcorp.org/browse/DM-35645) was resolved in November 2022, but I don't see a "final" list of checks either there or on [DM-37142](https://jira.lsstcorp.org/browse/DM-37142).

Reply:
It still remains to do, but I think the statement that we can have the pipeline raise on bad data is fine.

Reply:
My question was about "the proposed checks focus on poor-quality raw inputs; there are no checks specifically guarding DiaSource detection or the APDB" -- is that still true, or an overgeneralization?

Reply:
Yes, that's true as of today.